pyiron_atomistics/lammps/interactive.py

```diff
@@ -242,6 +243,84 @@ def interactive_initialize_interface(self):
         self._reset_interactive_run_command()
         self.interactive_structure_setter(self.structure)
 
+    def set_callback(self, function, n_call=1, n_apply=1, overload_internal_callback=False):
+        """
+        **********************
+        *** Expert feature ***
+        **********************
+
+        Set callback function that modifies forces.
+
+        Args:
+            function (function): User-defined function that returns forces (see below)
+            n_call (int): Make callback every `n_call` steps
+            n_apply (int): Apply callback forces every `n_apply` steps
+            overload_internal_callback (bool): Whether to overload internal callback (see below).
+                Overloading the internal callback function will have the advantage that the code
+                will not need to do expensive copying, BUT it is extremely error-prone. Make
+                sure that the code works without overloading the internal callback function first.
+
+        `function` must have the following form:
+
+        ```
+        def function(positions, ntimestep, nlocal):
+            your_evaluation
+            return forces
+        ```
+        where `positions` is the positions of all atoms on the local processor, `ntimestep` is the
+        current timestep and `nlocal` is the number of atoms on the current processor. `forces`
+        must be of the shape `(n_atoms, 3)`. The total translational force will be eliminated
+        inside pyiron.
+
+        If `overload_internal_callback` is set to `True`, then `function` must have the following
+        form:
+
+        ```
+        def function(ptr, timestep, nlocal, ids, x, fexternal):
+            your_evaluation
+        ```
+
+        with the following arguemnts:
+
+        - `ptr`: pointer provided by and simply passed back to external driver
+        - `timestep`: current LAMMPS timestep
+        - `nlocal`: # of atoms on this processor
+        - `ids`: list of atom IDs on this processor
+        - `x`: coordinates of atoms on this processor
+        - `fexternal`: forces to add to atoms on this processor
+
+        Note: Do NOT overwrite `fexternal`, because it points to the internal memory of LAMMPS and
+        therefore overwriting it will erase its functionality. E.g. DO `fexternal.fill(0)` and NOT
+        `fexternal = np.zeros_like(x)`.
+
+        Example: Add random forces
+
+        ```
+        from pyiron import Project
+
+        def random_forces(x, *args):
+            return 0.1 * np.random.randn(*x.shape)
+
+        pr = Project('RANDOM')
+        lmp = pr.create.job.Lammps('random_forces')
+        lmp.structure = your_structure
+        lmp.potential = your_potential
+        lmp.server.run_mode.interactive = True
+        lmp.set_callback(random_forces)
+        lmp.calc_md()
+        lmp.run()
+        lmp.interactive_close()
+        ```
+        """
```

Comment:
I guess you should also include something along the lines of example 2 in the docstring. If I understand correctly, it's this example that makes differences between `n_apply` and `n_call` relevant -- i.e. with a bare function they must always be the same, but with a class we can use `n_apply < n_call` and use the non-applied calls to build up some internal counter, e.g. using Welfords Online algo to track running means and exploit them in the intermittently applied force.

Reply:
yeah actually I don't know how things change if `n_apply` and `n_call` vary. I'm gonna check it this afternoon